File: python/rules/domain-modeling/no-nullable-constructor-defaults.py
from typing import Optional, Union

class AbstractTokenProvider:
    pass

class TokenService:
    # ruleid: python-no-nullable-constructor-defaults
    def __init__(self, token_provider: AbstractTokenProvider | None = None):
        self.token_provider = token_provider

class OptionalService:
    # ruleid: python-no-nullable-constructor-defaults
    def __init__(self, token_provider: Optional[AbstractTokenProvider] = None):
        self.token_provider = token_provider

class UnionService:
    # ruleid: python-no-nullable-constructor-defaults
    def __init__(self, token_provider: Union[AbstractTokenProvider, None] = None):
        self.token_provider = token_provider

class KeywordOnlyService:
    # ruleid: python-no-nullable-constructor-defaults
    def __init__(self, *, token_provider: AbstractTokenProvider | None = None):
        self.token_provider = token_provider

class UntypedService:
    # ruleid: python-no-nullable-constructor-defaults
    def __init__(self, token_provider = None):
        self.token_provider = token_provider

class ExplicitService:
    # ok: python-no-nullable-constructor-defaults
    def __init__(self, token_provider: AbstractTokenProvider):
        self.token_provider = token_provider

File: csharp/rules/domain-modeling/no-nullable-constructor-defaults.cs
namespace Thruput.Semgrep.Tests
{
    public interface ITokenProvider { }

    public class TokenService
    {
        private readonly ITokenProvider _provider;

        // ruleid: csharp-no-nullable-constructor-defaults
        public TokenService(ITokenProvider? tokenProvider = null)
        {
            _provider = tokenProvider;
        }
    }

    public class DefaultTokenService
    {
        private readonly ITokenProvider _provider;

        // ruleid: csharp-no-nullable-constructor-defaults
        public DefaultTokenService(ITokenProvider? tokenProvider = default)
        {
            _provider = tokenProvider;
        }
    }

    public class ExplicitTokenService
    {
        private readonly ITokenProvider _provider;

        // ok: csharp-no-nullable-constructor-defaults
        public ExplicitTokenService(ITokenProvider tokenProvider)
        {
            _provider = tokenProvider;
        }
    }
}
